modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToMessageDecoder;
+
+import java.util.List;
+
+@ChannelHandler.Sharable
+public class NettyByteBufSizer extends MessageToMessageDecoder<ByteBuf> {
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> out) {
+        final int maxFastWritableBytes = buf.maxFastWritableBytes();
+        final int readableBytes = buf.readableBytes();
+        // If max fast writeable bytes is greater than the number of readable bytes, this buffer is at least
+        // twice as big as necessary to contain the data. If that is the case, allocate a new buffer and
+        // copy.
+        int estimatedSize = maxFastWritableBytes + buf.writerIndex();
+        if (estimatedSize > 1024 && maxFastWritableBytes >= readableBytes) {
```

Comment:
Couldn't we, instead of rolling our own copying here, just use a call to `ByteBuf#capacity` and make Netty resize things? I.e. just do something like:

```
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> out) {
        final int readableBytes = buf.readableBytes();
        out.add(buf.capacity(readableBytes).retain());
    }
```

At least in some quick experimentation with the debugger that method seems to accomplish a similar if not the same thing we do here now but maybe I'm missing something?

Reply:
This seems fine. I guarded against less than 1024 because capacity will always copy small arrays which seems unnecessary.